### PR TITLE
Fix tuple types for 0.4

### PR DESCRIPTION
**Pull request by rened**
Continuing here from #228 to allow for collaborative editing.

### CODE REVIEW

**Comment by fundamental** [comment]
@rened if you need a hand getting this to work let me know where to start modifying as this issue is a blocker for me as I need to get at my data to start debugging/replicating a gc bug somewhere deep inside the base julia.

**Comment by rened** [comment]
@fundamental Thanks! I am waiting to see whether Jeff's recent fix now works for @simonster

File: src/jld_types.jl [hunk 1]
```diff
@@ -292,15 +328,16 @@ h5fieldtype{T,N}(parent::JldFile, ::Type{Array{T,N}}, ::Bool) = JLD_REF_TYPE
...
 else
-    h5fieldtype(parent::JldFile, T::(Type...), ::Bool) = JLD_REF_TYPE
+    h5fieldtype(parent::JldFile, T::TupleType, ::Bool) = JLD_REF_TYPE
 end
 
-function h5type(parent::JldFile, T::(Type...), commit::Bool)
-    !isa(T, (Union(Tuple, DataType)...)) && unknown_type_err(T)
-    T = T::(Union(Tuple, DataType)...)
+function h5type(parent::JldFile, T::TupleType, commit::Bool)
+    @assert isa(T, TupleType)
```
Comment (simonster) [inline src/jld_types.jl]: This is supposed to be a no-op, right?

**Comment by simonster** [comment]
Jeff's fix helped a bit, but I think the underlying issue is actually a dispatch bug that I need to track down...

File: src/jld_types.jl [hunk 2]
```diff
@@ -523,11 +560,12 @@ end
 ## Common functions for all non-special types (including gen_h5convert)
 
 # Whether this datatype should be stored as opaque
-isopaque(t::(Type...)) = isa(t, ())
-isopaque(t::DataType) = isempty(t.names)
+isopaque(t::TupleType) = t == EMPTY_TUPLE_TYPE
+# isopaque(t::DataType) = isempty(fieldnames(t))
+isopaque(t::DataType) = isa(t, TupleType) ? t == EMPTY_TUPLE_TYPE : isempty(fieldnames(t))
```
Comment (simonster) [inline src/jld_types.jl]: This is how I noticed JuliaLang/julia#10978, but I think `isopaque(t::TupleType)` should be called if `isa(t, TupleType)`, and not this method.

**Comment by simonster** [comment]
Filed https://github.com/JuliaLang/julia/issues/10995 with test case

**Comment by simonster** [comment]
Field https://github.com/JuliaLang/julia/issues/10998. This one I think I can work around.

**Comment by sebastiang** [comment]
Wish I'd looked earlier to see this work going on! Was hacking away at it a bit myself. Let me know if you need any help. I just got to the `SimpleVector` issue myself, but perhaps you have a workaround? It caused me to file https://github.com/JuliaLang/julia/issues/10999

**Comment by simonster** [comment]
The SimpleVector issue is not really a bug, just an implementation change. It was among the easier things to handle.

The PR here currently passes the test suite (although not the JLDArchives tests yet). However, you probably don't want to use it yet because I can't guarantee that any files you save with it will be readable by when this is in a state to be merged to master. Anything not involving tuples or types with fields that are tuples is probably okay, but anything else might not work. I also need to add more backwards compatibility tests to JLDArchives to see how badly this breaks reading older files, and fix that as best as I can.

**Comment by sebastiang** [comment]
Understood! I have an aversion to magical reflection based serialization in any language, so was just planning on using the raw HDF5 interfaces mostly anyway.

**Comment by timholy** [comment]
Thanks a bunch for this, everyone! I'll also link https://github.com/JuliaLang/julia/issues/10999#issuecomment-96158871 as a general issue of growing concern to me (but without me yet having carved out the time to find a general solution solution).

**Comment by simonster** [comment]
I added test files to JLDArchives that have our entire test suite saved with 0.3 and master. Those tests all pass (files written with both versions can be read by both versions), except for the SubArray tests, which fail because the structure of the type has changed. (Maybe we should reconstruct the type in this case too instead of throwing an error.)

This doesn't mean everything works, because covering all our code paths isn't actually sufficient to make sure that all old files can be read by newer Julia, but at least the situation is not too bad.

**Comment by simonster** [comment]
I can't figure out what's going on with the release build on OS X, which is intermittently failing with a completely empty log (??), but I suspect that's not related to the package at all. Other than that I think this is ready to merge.

File: src/jld_types.jl (hunk 2 above)
Comment (timholy) [inline src/jld_types.jl]: Do we still need this?

File: src/jld_types.jl [hunk 3]
```diff
@@ -631,6 +659,9 @@ function jldatatype(parent::JldFile, dtype::HDF5Datatype)
         if T == UnsupportedType
             warn("type $typename not present in workspace; reconstructing")
             T = reconstruct_type(parent, dtype, typename)
+        elseif T == Type
+            # https://github.com/JuliaLang/julia/issues/10998
```
Comment (timholy) [inline src/jld_types.jl]: Likewise here?

File: src/plain.jl [hunk 4]
```diff
@@ -431,10 +431,8 @@ end
 type EmptyArray{T}; end
 
 # Stub types to encode fixed-size arrays for H5T_ARRAY
-immutable DimSize{N}; end  # Int-wrapper (can't use tuple of Int as param)
-immutable FixedArray{T,D<:(DimSize...)}; end
-dimsize{N}(::Type{DimSize{N}}) = N
-size{T,D}(::Type{FixedArray{T,D}}) = map(dimsize, D)::(Int...)
+immutable FixedArray{T,D}; end
+size{T,D}(::Type{FixedArray{T,D}}) = D
```
Comment (timholy) [inline src/plain.jl]: I had forgotten that this capability had gotten backported to 0.3. Nice!

**Comment by timholy** [comment]
It all looks great. Thanks to both of you! Merge at will.

File: src/jld_types.jl (hunk 2 above)
Comment (simonster) [inline src/jld_types.jl]: The tests fail without it, so the answer seems to be yes, although I'm not sure why.

File: src/jld_types.jl (hunk 3 above)
Comment (simonster) [inline src/jld_types.jl]: Probably not. I'll check.

**Comment by rened** [comment]
Thanks to you two for tackling the hard parts!

**Comment by timholy** [comment]
I didn't do anything, you two deserve the thanks!

**Pull request merged**